### PR TITLE
Kfbsd v3 Code working for ARM FreeBSD Kernel.

### PR DESCRIPTION
**Pull request by stelkasouridis**
Several modifications were made to work with ARM FreeBSD Kernel. The code is tested and works successfully with a FreeBSD Kernel module.

Only required steps are:

sudo mkdir bld
cd bld
sudo ../configure --host=arm-kfbsd
sudo make sqlite3-all.c

### CODE REVIEW

File: src/date.c [hunk 1]
```diff
@@ -1563,6 +1568,48 @@ static void ctimestampFunc(
...
+  sqlite3_mutex_leave(sqlite3MutexAlloc(SQLITE_MUTEX_STATIC_MAIN));
+#endif
+  if( pTm ){
+    strftime(zBuf, 20, zFormat, &sNow);
+    sqlite3_result_text(context, zBuf, -1, SQLITE_TRANSIENT);
+  }
+  */
+  zBuf[0] = '0';
+  for (int i = 1; i < 20; i++) zBuf[i] = zBuf[i-1];
+  sqlite3_result_text(context, zBuf, -1, SQLITE_TRANSIENT);
```
Comment (robertsoule) [inline src/date.c:1610]: @stelkasouridis : while I think it is okay to stub out functionality for now, I am worried that this has a potential to fail silently. I think you need to add some code here to either emit trace that this should not be called or to initiate some type of failure/assertion.

File: src/expr.c [hunk 2]
```diff
@@ -4450,7 +4450,11 @@ int sqlite3ExprCodeTarget(Parse *pParse, Expr *pExpr, int target){
       }
       pCol = &pAggInfo->aCol[pExpr->iAgg];
       if( !pAggInfo->directMode ){
+        #ifdef FREEBSD_KERNEL
+        //todo: STELIOS
```
Comment (robertsoule) [inline src/expr.c:4455]: @stelkasouridis : It seems like you should emit some warning here that the code isn't ready yet.

File: src/func.c [hunk 3]
```diff
@@ -1769,11 +1769,15 @@ static void kahanBabuskaNeumaierStep(
 ){
   volatile double s = pSum->rSum;
   volatile double t = s + r;
+  #ifdef FREEBSD_KERNEL
+  //todo: STELIOS
```
Comment (robertsoule) [inline src/func.c:1773]: @stelkasouridis : should you also emit a warning of fail if this code path is traversed?

File: src/os.c [hunk 4]
```diff
@@ -374,7 +382,11 @@ sqlite3_vfs *sqlite3_vfs_find(const char *zVfs){
   sqlite3_mutex_enter(mutex);
   for(pVfs = vfsList; pVfs; pVfs=pVfs->pNext){
     if( zVfs==0 ) break;
+    #ifdef FREEBSD_KERNEL
+    //todo: STELIOS
```
Comment (robertsoule) [inline src/os.c:387]: Emit warning?

File: src/os_kfbsd.c [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+#ifdef FREEBSD_KERNEL
+
+#include "os_kfbsd.h"
+
+/*
+** Initialize the operating system interface.
+*/
+int sqlite3_os_init(void){
+  /* TODO: Implement FreeBSD-specific initialization here. */
```
Comment (robertsoule) [inline src/os_kfbsd.c:15]: What functionality should go here? Should you emit a warning that this is not yet done? Why don't you need to have something here in order for your demo to work?

File: src/os_kfbsd.c [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+** Deinitialize the operating system interface.
+*/
+int sqlite3_os_end(void){
+  /* TODO: Implement FreeBSD-specific deinitialization here. */
+  return SQLITE_OK;
+}
+
+
+static int kern_vfs_open(sqlite3_vfs *vfs, sqlite3_filename zName, sqlite3_file *file,
+                         int flags, int *pOutFlags) {
```
Comment (robertsoule) [inline src/os_kfbsd.c:31]: Fix the formatting?

File: src/os_kfbsd.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
+#ifndef FREEBSD_KERNEL_H
+#define FREEBSD_KERNEL_H
+
+double __floatdidf(long a) {
+    //todo: STELIOS
```
Comment (robertsoule) [inline src/os_kfbsd.h:5]: @stelkasouridis : can you please emit a warning if these methods are called? I am worried that they are going to fail silently.

**Review by robertsoule — CHANGES_REQUESTED** [review]
Stelios, can you please add code to emit warnings for code paths where you have not yet implemented the logic? If we don't have some sort of indication, then we might falsely assume that the code is working.

File: ext/misc/amatch.c [hunk 8]
```diff
@@ -1405,7 +1405,7 @@ static int amatchBestIndex(
     pIdxInfo->orderByConsumed = 1;
   }
   pIdxInfo->estimatedCost = (double)10000;
-
```
Comment (robertsoule) [inline ext/misc/amatch.c]: @stelkasouridis : can you please revert to the committed version of this code? It looks like you removed a newline, and we don't really want to commit that change.

File: ext/misc/prefixes.c [hunk 9]
```diff
@@ -204,7 +204,7 @@ static int prefixesBestIndex(
   ** (original_string) and use it if at all possible */
   int i;
   const struct sqlite3_index_constraint *p;
-
+
```
Comment (robertsoule) [inline ext/misc/prefixes.c]: @stelkasouridis : please revert to committed version of this file. We don't want to commit changes that just include whitespace changes.

File: src/date.c (hunk 1 above)
Comment (stelkasouridis) [reply src/date.c:1610]: Emitted trace that this function should not be called.

Comment (robertsoule) [reply src/date.c:1610]: Maybe "currentTimeFunc unimplemented on FreeBSD kernel" is more informative?

File: ext/lsm1/lsm-test/lsmtest_main.c
Comment (gvnn3) [inline ext/lsm1/lsm-test/lsmtest_main.c:1]: Seems like this was a whitespace change that you don't need.  Please remove it from the PR.

File: sqlite_cfg.h.in
Comment (gvnn3) [inline sqlite_cfg.h.in:1]: Why are lines being removed from this file?  In theory we should only see additions.

File: src/os.c
Comment (gvnn3) [inline src/os.c:1]: I'm a little confused on what this function is supposed to do normally.  Is it just to get a random number or to seed a random number generator?  The kernel has a source of randomness that can be used if it's required here.

**Review by gvnn3 — CHANGES_REQUESTED** [review]

File: configure.ac [hunk 10]
```diff
@@ -651,10 +698,17 @@ fi
 AC_SUBST(USE_AMALGAMATION)
 
 #########
-# Look for zlib.  Only needed by extensions and by the sqlite3.exe shell
-AC_CHECK_HEADERS(zlib.h)
-AC_SEARCH_LIBS(deflate, z, [HAVE_ZLIB="-DSQLITE_HAVE_ZLIB=1"], [HAVE_ZLIB=""])
-AC_SUBST(HAVE_ZLIB)
```
Comment (robertsoule) [inline configure.ac:657]: Why are these lines removed?

File: sqlite_cfg.h.in [hunk 11]
```diff
@@ -42,9 +48,6 @@
 /* Define to 1 if you have the `malloc_usable_size' function. */
 #undef HAVE_MALLOC_USABLE_SIZE
 
-/* Define to 1 if you have the <memory.h> header file. */
-#undef HAVE_MEMORY_H
-
```
Comment (robertsoule) [inline sqlite_cfg.h.in:47]: Why is this removed?

File: sqlite_cfg.h.in [hunk 12]
```diff
@@ -127,16 +132,28 @@
...
 
-/* Define to 1 if you have the ANSI C header files. */
-#undef STDC_HEADERS
+/* FreeBSD Kernel uses no datetime functions */
+#undef SQLITE_OMIT_DATETIME_FUNCS
 
-/* Enable large inode numbers on Mac OS X 10.5.  */
-#ifndef _DARWIN_USE_64_BIT_INODE
-# define _DARWIN_USE_64_BIT_INODE 1
-#endif
```
Comment (robertsoule) [inline sqlite_cfg.h.in:136]: Why are these lines removed?

**Review by robertsoule — APPROVED** [review]

**Review by gvnn3 — APPROVED** [review]

**Pull request closed without merging**